src/js/view/controls/components/menu/menu.js

```diff
@@ -0,0 +1,499 @@
+import UI from 'utils/ui';
+import Events from 'utils/backbone.events';
+import { 
+    createElement, 
+    emptyElement, 
+    prependChild, 
+    nextSibling, 
+    previousSibling
+} from 'utils/dom';
+import button from 'view/controls/components/button';
+import { cloneIcon } from 'view/controls/icons';
+import { MenuItem, RadioMenuItem } from 'view/controls/components/menu/menu-item';
+import { MenuTemplate } from 'view/controls/templates/menu/menu';
+import categoryButton from 'view/controls/components/menu/category-button';
+import { isRtl } from 'utils/language';
+
+export default class Menu extends Events {
+    constructor(_name, _parentMenu, _localization, _template = MenuTemplate) {
+        super();
+        this.name = _name;
+        this.isSubmenu = !!_parentMenu;
+        this.el = createElement(_template(this.isSubmenu, _name));
+        this.topbar = this.el.querySelector(`.jw-${this.name}-topbar`);
+        this.children = {};
+        this.openMenus = [];
+        this.items = [];
+        this.visible = false;
+        this.parentMenu = _parentMenu;
+        this.mainMenu = !this.parentMenu ? this : this.parentMenu.mainMenu;
+        this.closeButton = (this.parentMenu && this.parentMenu.closeButton) || this.createCloseButton(_localization);
+        this.categoryButton = null;
+        if (this.isSubmenu) {
+            if (this.parentMenu.name === 'settings') {
+                this.createCategoryButton(_localization);
+            }
+            this.parentMenu.appendMenu(this);
+        } else {
+            this.ui = addGlobalMenuKeyListener(this);
+        }
+        this.onDocumentClick = this.onDocumentClick.bind(this);
+        this.open = this.open.bind(this);
+        this.close = this.close.bind(this);
+        this.toggle = this.toggle.bind(this);
+    }
+    createCloseButton(localization) {
+        const closeButton = button('jw-settings-close', (e) => this.close(e), localization.close, [cloneIcon('close')]);
+        this.topbar.appendChild(closeButton.element());
+        closeButton.show();
+        closeButton.ui.on('keydown', (e) => {
+            const sourceEvent = e.sourceEvent;
+            const key = sourceEvent.key.replace(/(Arrow|ape)/, '');
+            // Close settings menu when enter is pressed on the close button
+            // or when tab or right arrow key is pressed since it is the last element in topbar
+            if (key === 'Enter' || key === 'Right' || (key === 'Tab' && !sourceEvent.shiftKey)) {
```

Comment:
does `tab` not get covered by usual focus/blur functionality?

Reply:
We do a lot of custom navigation logic to conform to W3C accessibility standards. So there's a default behavior, but we want to overwrite that with the expected behavior. Refactoring the menu to have this behavior via HTML alone may be possible, but is outside of the scope of this refactor.